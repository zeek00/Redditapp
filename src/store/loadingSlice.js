import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    loading: false,
    completed: false,
    error:false

}

const loadingSlice = createSlice({ 
    name: "loadingState",
    initialState,
    reducers: {
        changeLoadingState(state,action){
            state.loading = true
            state.completed = false
            state.error = false
        },
        changeCompletedState(state,action){
            state.loading = false
            state.completed = true
            state.error = false
        },
        changeErrorState(state,action){
            state.loading = false
            state.completed = false
            state.error = true
        },
        changeToInitialState(state,action){
            state.loading = false
            state.completed = false
            state.error = false
        },
    }
})


export const {changeLoadingState,  changeCompletedState,changeErrorState, changeToInitialState } = loadingSlice.actions

export default loadingSlice.reducer