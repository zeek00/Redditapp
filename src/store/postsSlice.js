import { createSlice } from "@reduxjs/toolkit";

const initialState = {
    popularPosts: [],
    bestPosts: [],
    currSearchPosts: []

   
}

const options = {
    name: "postsData",
    initialState,
    reducers: {
        addPopularPosts(state,action){
            state.popularPosts = action.payload
        },
        addBestPosts(state,action){
            state.bestPosts = action.payload
        },
        addCurrSearchPosts(state,action){
            state.currSearchPosts = action.payload
        }
    }
}

const postsSlice = createSlice(options)

export const { addPopularPosts, addBestPosts } = postsSlice.actions
export default postsSlice.reducer