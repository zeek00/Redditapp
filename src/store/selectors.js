export const loadingSelector = state => state.loadingReducer.loading
export const popularPostsSelector = state => state.postsReducer.popularPosts