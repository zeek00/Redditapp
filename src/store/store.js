import { configureStore } from "@reduxjs/toolkit";
import postsSliceReducer from "./postsSlice"
import loadingSliceReducer from "./loadingSlice";

const store = configureStore({
    reducer: {
        postsReducer : postsSliceReducer,
        loadingReducer: loadingSliceReducer
    },
});

export default store