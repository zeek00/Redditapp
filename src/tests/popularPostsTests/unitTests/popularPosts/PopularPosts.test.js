import React from 'react';
import {screen, act} from '@testing-library/react'
import '@testing-library/jest-dom'
import {renderWithProviders} from '../../../storeForTests/loadingAndPostReducersStore';
import {changeLoadingState} from "../../../../store/loadingSlice";
import 'core-js/stable/structured-clone';

const PostComponent = require("../../../../components/Posts/Post")
const popular = require("../../../../api/popularPosts/getPopularPosts")
const popularPostsData = require("./PopularPostsData")

describe("PopularPosts", () => {
    afterEach(() => {
        jest.restoreAllMocks();
      });
    


  it("getPopularPosts is loading data, then Loading option comes up", async () => {

        const receivePopularPostsSpy = jest.spyOn(popular, "getPopularPosts")
            .mockImplementationOnce((dispatch) => {
                console.log('receivePopularPostsSpy called');
                dispatch(changeLoadingState({message: "to loading"}))

                return []
            })

        const PostsComponentSpy = jest.spyOn(PostComponent, "Post")
            .mockImplementationOnce((calledWith, p, y) => {
                console.log('PostsComponentSpy1 called' + JSON.stringify(calledWith));
                return ;
            })

        const PopularPosts = (await import("../../../../api/popularPosts/PopularPosts")).default
        const utils = renderWithProviders(<PopularPosts/>)

        await act(async () => {
            // eslint-disable-next-line testing-library/no-debugging-utils
            await screen.debug()

        })

        expect(await receivePopularPostsSpy).toHaveBeenCalled()
        expect(PostsComponentSpy).toHaveBeenCalledTimes(0)
        expect(utils).toMatchSnapshot();


    })



    it("received null from getSortedPopularPosts to display", async () => {

        const receivePopularPostsSpy = jest.spyOn(popular, "getPopularPosts")
            .mockImplementationOnce(() => {
                console.log('receivePopularPostsSpy called');
                return null
            })

        const PostsComponentSpy = jest.spyOn(PostComponent, "Post")
            .mockImplementationOnce((calledWith, p, y) => {
                console.log('PostsComponentSpy1 called' + JSON.stringify(calledWith));
                return ;
            })

        const PopularPosts = (await import("../../../../api/popularPosts/PopularPosts")).default
        const utils = renderWithProviders(<PopularPosts/>)

        await act(async () => {
            // eslint-disable-next-line testing-library/no-debugging-utils
            await screen.debug()

        })

        expect(await receivePopularPostsSpy).toHaveBeenCalled()
        expect(PostsComponentSpy).toHaveBeenCalledTimes(0)
        expect(utils).toMatchSnapshot();


    })



it("received empty [] of posts to display", async () => {

        const receivePopularPostsSpy = jest.spyOn(popular, "getPopularPosts")
            .mockImplementationOnce(() => {
                console.log('receivePopularPostsSpy called');
                return []
            })

        const PostsComponentSpy = jest.spyOn(PostComponent, "Post")
            .mockImplementation((calledWith, p, y) => {
                console.log('PostsComponentSpy1 called' + JSON.stringify(calledWith));
                return ;
            })

        const PopularPosts = (await import("../../../../api/popularPosts/PopularPosts")).default
        const utils = renderWithProviders(<PopularPosts/>)

        await act(async () => {
            // eslint-disable-next-line testing-library/no-debugging-utils
            await screen.debug()

        })

        expect(await receivePopularPostsSpy).toHaveBeenCalledTimes(1)
        expect(PostsComponentSpy).toHaveBeenCalledTimes(0)
        expect(utils).toMatchSnapshot();


    })



  it("received two valid posts to display", async () => {

        const receivePopularPostsSpy = jest.spyOn(popular, "getPopularPosts")
            .mockImplementationOnce(() => {
                console.log('receivePopularPostsSpy called');
                const twoValidPosts = structuredClone(popularPostsData.PopularPostsData.twoSortedValidPosts)
                return twoValidPosts
            })

        const PostsComponentSpy = jest.spyOn(PostComponent, "Post")
            .mockImplementationOnce((calledWith, p, y) => {
                console.log('PostsComponentSpy1 called' + JSON.stringify(calledWith));
                return <p>PostsComponentSpy1</p>;
            })
            .mockImplementationOnce((calledWith, p, y) => {
                console.log('PostsComponentSpy2 called' + JSON.stringify(calledWith));
                return <p>PostsComponentSpy2</p>;
            })


        const PopularPosts = (await import("../../../../api/popularPosts/PopularPosts")).default
        const utils = renderWithProviders(<PopularPosts/>)

        await act(async () => {

            // eslint-disable-next-line testing-library/no-debugging-utils
            await screen.debug()

        })

        expect(await receivePopularPostsSpy).toHaveBeenCalled()
        expect(PostsComponentSpy).toHaveBeenCalledTimes(2)
        expect(PostsComponentSpy).toHaveBeenNthCalledWith(1, {onPost: popularPostsData.PopularPostsData.twoSortedValidPosts[0]}, {})
        expect(PostsComponentSpy).toHaveBeenNthCalledWith(2, {onPost: popularPostsData.PopularPostsData.twoSortedValidPosts[1]}, {})
        expect(utils).toMatchSnapshot();


    })




})




