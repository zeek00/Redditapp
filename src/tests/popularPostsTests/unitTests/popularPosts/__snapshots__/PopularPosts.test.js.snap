// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopularPosts getPopularPosts is loading data, then Loading option comes up 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="div"
      >
        <p>
           Loading....
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="div"
    >
      <p>
         Loading....
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`PopularPosts received empty [] of posts to display 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="div"
      >
        <ul
          class="ul"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="div"
    >
      <ul
        class="ul"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`PopularPosts received null from getSortedPopularPosts to display 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="div"
      >
        <ul
          class="ul"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="div"
    >
      <ul
        class="ul"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;

exports[`PopularPosts received two valid posts to display 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="div"
      >
        <ul
          class="ul"
        >
          <li>
            <p>
              PostsComponentSpy1
            </p>
          </li>
          <li>
            <p>
              PostsComponentSpy2
            </p>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="div"
    >
      <ul
        class="ul"
      >
        <li>
          <p>
            PostsComponentSpy1
          </p>
        </li>
        <li>
          <p>
            PostsComponentSpy2
          </p>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
